### ISSUE TITLE
DNN test

### ISSUE THREAD

**Issue by TranTony** [TranTony/DeepLearningBooks#1]
from __future__ import division, print_function, absolute_import

import tensorflow as tf
import numpy as np
import matplotlib.pyplot as plt

from array import array

TRAIN_SET = "lst/DNN.lst"
TO_EXTRACT = "data/data.lst"
SPRO_FEATURES_PATH = "data/prm_dnn/"
SPRO_FEATURES_EXT = ".spro"
DNN_FEATURES_PATH = "data/prm_dnn/"
DNN_FEATURES_EXT = ".dnn"

# Training Parameters
LEARNING_RATE = 0.01
NUM_STEPS = 5000
BATCH_SIZE = 256
DISPLAY_STEP = 1000

# Network Parameters
NUM_INPUT = 60
NORMALIZE_INPUT = True  # in range [0..1]

# https://stackoverflow.com/questions/40994583/how-to-implement-tensorflows-next-batch-for-own-data
def next_batch(num, data, labels):
	'''
	Return a total of `num` random samples and labels. 
	'''
	idx = np.arange(0 , len(data))
	np.random.shuffle(idx)
	idx = idx[:num]
	data_shuffle = [data[ i] for i in idx]
	labels_shuffle = [labels[ i] for i in idx]
	return np.asarray(data_shuffle), np.asarray(labels_shuffle)

# Normalize num into [0..1] range
# Vectorized and scalar versions
def normalize01(num, max_val, min_val):
	return (float(num) - min_val)/(max_val - min_val)
normalize01_vect = np.vectorize(normalize01)

# Read training data
train_mfccs = []
with open(TRAIN_SET) as f1:
	for l1 in f1.readlines():
		filename = l1.strip()
		with open(SPRO_FEATURES_PATH + filename + SPRO_FEATURES_EXT) as f2:
			for l2 in f2.readlines():
				coefs = l2.strip().split(" ")
				coefs_fl = [float(c) for c in coefs]
				if NORMALIZE_INPUT:
					coefs_fl = normalize01_vect(coefs_fl, max(coefs_fl), min(coefs_fl))
				train_mfccs.append(coefs_fl)
print("Done reading training data")

# tf Graph input
X = tf.placeholder("float", [None, NUM_INPUT])

############ MULTI LAYERS
NUM_HIDDEN_1 = 30
NUM_HIDDEN_2 = 32
NUM_HIDDEN_3 = 32

weights = {
    'encoder_h1': tf.Variable(tf.random_normal([NUM_INPUT, NUM_HIDDEN_1])),
    'encoder_h2': tf.Variable(tf.random_normal([NUM_HIDDEN_1, NUM_HIDDEN_2])),
    'encoder_h3': tf.Variable(tf.random_normal([NUM_HIDDEN_2, NUM_HIDDEN_3])),
    'decoder_h3': tf.Variable(tf.random_normal([NUM_HIDDEN_3, NUM_HIDDEN_2])),
    'decoder_h2': tf.Variable(tf.random_normal([NUM_HIDDEN_2, NUM_HIDDEN_1])),
    'decoder_h1': tf.Variable(tf.random_normal([NUM_HIDDEN_1, NUM_INPUT])),
}
biases = {
    'encoder_b1': tf.Variable(tf.random_normal([NUM_HIDDEN_1])),
    'encoder_b2': tf.Variable(tf.random_normal([NUM_HIDDEN_2])),
    'encoder_b3': tf.Variable(tf.random_normal([NUM_HIDDEN_3])),
    'decoder_b3': tf.Variable(tf.random_normal([NUM_HIDDEN_2])),
    'decoder_b2': tf.Variable(tf.random_normal([NUM_HIDDEN_1])),
    'decoder_b1': tf.Variable(tf.random_normal([NUM_INPUT])),
}

# Building the encoder
def encoder(x):
    layer_1 = tf.nn.sigmoid(tf.add(tf.matmul(x, weights['encoder_h1']),
                                   biases['encoder_b1']))
#     layer_2 = tf.nn.sigmoid(tf.add(tf.matmul(layer_1, weights['encoder_h2']),
#                                    biases['encoder_b2']))
#     layer_3 = tf.nn.sigmoid(tf.add(tf.matmul(layer_2, weights['encoder_h3']),
#                                    biases['encoder_b3']))
    return layer_1

# Building the decoder
def decoder(x):
#     layer_3 = tf.nn.sigmoid(tf.add(tf.matmul(x, weights['decoder_h3']),
#                                    biases['decoder_b3']))
#     layer_2 = tf.nn.sigmoid(tf.add(tf.matmul(layer_3, weights['decoder_h2']),
#                                    biases['decoder_b2']))
    layer_1 = tf.nn.sigmoid(tf.add(tf.matmul(x, weights['decoder_h1']),
                                   biases['decoder_b1']))
    return layer_1

# Construct model
encoder_op = encoder(X)
decoder_op = decoder(encoder_op)

# Prediction
y_pred = decoder_op
# Targets (Labels) are the input data.
y_true = X

# Define loss and optimizer, minimize the squared error
loss = tf.reduce_mean(tf.pow(y_true - y_pred, 2))
optimizer = tf.train.RMSPropOptimizer(LEARNING_RATE).minimize(loss)

# Initialize the variables (i.e. assign their default value)
init = tf.global_variables_initializer()

# Start Training
# Start a new TF session
with tf.Session() as sess:

	# Run the initializer
	sess.run(init)

	# Training
	for i in range(1, NUM_STEPS+1):
		# Prepare Data
		# Get the next batch of data
		batch_x, _ = next_batch(BATCH_SIZE, train_mfccs, train_mfccs)

		# Run optimization op (backprop) and cost op (to get loss value)
		_, l = sess.run([optimizer, loss], feed_dict={X: batch_x})
		# Display logs per step
		if i % DISPLAY_STEP == 0 or i == 1:
			print('Step %i: Minibatch Loss: %f' % (i, l))
	print('Done training')
	
	# Extract middle layer features
	with open(TO_EXTRACT) as f1:
		for l1 in f1.readlines():
			filename = l1.strip()
			outfile_bin = open(DNN_FEATURES_PATH + filename + DNN_FEATURES_EXT, "w")
			outfile_txt = open(DNN_FEATURES_PATH + filename + DNN_FEATURES_EXT + ".txt", "w")
			buff = []
			with open(SPRO_FEATURES_PATH + filename + SPRO_FEATURES_EXT) as f2:
				for l2 in f2.readlines():
					coefs_fl = [float(c) for c in l2.strip().split(" ")]
					if NORMALIZE_INPUT:
						coefs_fl = normalize01_vect(coefs_fl, max(coefs_fl), min(coefs_fl))
					features = sess.run(encoder_op, feed_dict={X: [coefs_fl]})
					buff.extend(features[0][:])
					outfile_txt.write(' '.join([str(c) for c in features[0]]) + '\n')
# 					features_output = sess.run(decoder_op, feed_dict={X: [coefs]})
			buff_out = array('f', buff)
			buff_out.tofile(outfile_bin)
			outfile_bin.close() 
			outfile_txt.close()
	print("Done extracting")
